src/object/srv_ec_aggregate.c

```diff
@@ -1419,46 +1419,41 @@ agg_process_holes_ult(void *arg)
 	unsigned int		 k = ec_age2k(entry);
 	unsigned long		 ss = entry->ae_cur_stripe.as_stripenum *
 					k * len;
-	unsigned int		 last_ext_end = 0;
+	unsigned long		 last_ext_end = 0;
 	unsigned int		 ext_cnt = 0;
 	unsigned int		 ext_tot_len = 0;
 	unsigned int		 failed_tgts_cnt = 0;
 	int			 rc = 0;
 
-	D_ALLOC_ARRAY(stripe_ud->asu_recxs,
-		      entry->ae_cur_stripe.as_extent_cnt + 1);
-	if (stripe_ud->asu_recxs == NULL) {
-		rc = -DER_NOMEM;
-		goto out;
-	}
 	/* Process extent list to find what to re-replicate -- build recx array
 	 */
 	d_list_for_each_entry_safe(agg_extent, ext_tmp,
-				   &entry->ae_cur_stripe.as_dextents,
-				   ae_link) {
-		if (agg_extent->ae_hole ||
-		    agg_extent->ae_epoch < entry->ae_par_extent.ape_epoch)
+				   &entry->ae_cur_stripe.as_dextents, ae_link) {
+		if (agg_extent->ae_epoch < entry->ae_par_extent.ape_epoch)
 			continue;
 		if (agg_extent->ae_recx.rx_idx - ss > last_ext_end) {
 			stripe_ud->asu_recxs[ext_cnt].rx_idx =
 				ss + last_ext_end;
 			stripe_ud->asu_recxs[ext_cnt].rx_nr =
-				agg_extent->ae_recx.rx_idx -
-				ss - last_ext_end;
+				agg_extent->ae_recx.rx_idx - ss -
+				last_ext_end;
 			ext_tot_len +=
 			stripe_ud->asu_recxs[ext_cnt++].rx_nr;
 		}
-		last_ext_end += agg_extent->ae_recx.rx_idx +
-				agg_extent->ae_recx.rx_nr - ss;
+		last_ext_end = agg_extent->ae_recx.rx_idx +
+			agg_extent->ae_recx.rx_nr - ss;
+		if (last_ext_end >= ss + k * len)
+			break;
 	}
 
 	if (last_ext_end < k * len) {
```

Comment:
what if no hole extent higher than parity epoch found by above for(...), because "as_has_holes" flag was set if with hole ext (despite of its epoch is higher or lower than parity epoch). That case will get last_ext_end == 0, but need not replicate any exts, no?

Reply:
Correct.  Only having holes older than parity means don't re-replicate.  This iteration of the extents is first place this can be determined, so I'll need to change the code.  It might be better to do this iteration in the function that calls this ULT, instead of waiting until the communication ULT is forked.